opsimulator/opsimulator_test.go

```diff
@@ -0,0 +1,197 @@
+package opsimulator
+
+import (
+	"testing"
+
+	"github.com/ethereum-optimism/supersim/config"
+	"github.com/ethereum-optimism/supersim/testutils"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/stretchr/testify/require"
+)
+
+// MockChainWithDependencySet extends MockChain to support dependency set configuration
+type MockChainWithDependencySet struct {
+	*testutils.MockChain
+	chainID       uint64
+	dependencySet []uint64
+}
+
+func NewMockChainWithDependencySet(chainID uint64, dependencySet []uint64) *MockChainWithDependencySet {
+	return &MockChainWithDependencySet{
+		MockChain:     testutils.NewMockChain(),
+		chainID:       chainID,
+		dependencySet: dependencySet,
+	}
+}
+
+func (c *MockChainWithDependencySet) Config() *config.ChainConfig {
+	return &config.ChainConfig{
+		Name:      "mockchain",
+		ChainID:   c.chainID,
+		BlockTime: 2,
+		L2Config: &config.L2Config{
+			L1ChainID:     900,
+			DependencySet: c.dependencySet,
+		},
+	}
+}
+
+func TestCheckInteropInvariants_DependencySetValidation(t *testing.T) {
+	tests := []struct {
+		name                   string
+		executingChainID       uint64
+		executingDependencySet []uint64
+		sourceChainID          uint64
+		expectError            bool
+		expectedErrorContains  string
+	}{
+		{
+			name:                   "Valid: Source chain in dependency set",
+			executingChainID:       901,
+			executingDependencySet: []uint64{902, 903},
+			sourceChainID:          902,
+			expectError:            false,
+		},
+		{
+			name:                   "Invalid: Source chain not in dependency set",
+			executingChainID:       901,
+			executingDependencySet: []uint64{902},
+			sourceChainID:          903,
+			expectError:            true,
+			expectedErrorContains:  "executing message in block (chain 901) may not execute message from chain 903: not in dependency set",
+		},
+		{
+			name:                   "Invalid: Empty dependency set",
+			executingChainID:       903,
+			executingDependencySet: []uint64{},
+			sourceChainID:          901,
+			expectError:            true,
+			expectedErrorContains:  "executing message in block (chain 903) may not execute message from chain 901: not in dependency set",
+		},
+		{
+			name:                   "Valid: Multiple chains in dependency set",
+			executingChainID:       903,
+			executingDependencySet: []uint64{901, 902},
+			sourceChainID:          901,
+			expectError:            false,
+		},
+		{
+			name:                   "Valid: Same-chain message (source == destination)",
+			executingChainID:       901,
+			executingDependencySet: []uint64{902}, // 901 is NOT in its own dependency set
```

Comment:
nit: this comment confused me a little bit since even though 901 is not in its own dependency set, it is still a valid message. it could be good to remove it, unless you think it is helpful.

Reply:
makes sense, removed.